### PR TITLE
Include headers explicitly and add constructors to comply C++20

### PR DESCRIPTION
**Pull request by HJLebbink**
The constructor update is necessary for the project to compile in C++20 due to struct initialization rules made more strict in C++20. The header update is to reduce the implementations details (ic header includes) leak.

### CODE REVIEW

File: examples/PutObject.cc [hunk 1]
```diff
@@ -13,7 +13,16 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <fstream>
```
Comment (balamurugana) [inline examples/PutObject.cc:16]: is this explicit declaration a standard or best practice?

File: .gitattributes [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Set the default behavior, in case people don't have core.autocrlf set.
+* text=auto
+
+# Declare files that will always have LF line endings on checkout.
+*.sh test eol=lf
```
Comment (balamurugana) [inline .gitattributes:5]: missing newline.

**Review by balamurugana — APPROVED** [review]

**Comment by kobalicek** [comment]
After thinking about this I would probably change the pattern used in constructors to the following:

```
class X {
  some_type _a;
  some_type _b;

  X(some_type a, some_type b)
    _a(std::move(a)),
    _b(std::move(b)) {}
};
```

The reason is that with this design the user can actually decide which argument to move and which to copy. The copy is guaranteed to only happen once, so there is only the extra move per value, which seems okay to me.

**Comment by piotr-topnotch** [comment]
> After thinking about this I would probably change the pattern used in constructors to the following:

I concur with this recommendation as being the least intrusive.

**Review by piotr-topnotch — CHANGES_REQUESTED** [review]

**Pull request merged**